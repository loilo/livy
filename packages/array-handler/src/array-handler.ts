import type {
  LogRecord,
  ResettableInterface,
  SyncHandlerInterface,
} from '@livy/contracts'
import { AbstractLevelBubbleHandler } from '@livy/util/handlers/abstract-level-bubble-handler'
import { MirrorSyncHandlerMixin } from '@livy/util/handlers/mirror-sync-handler-mixin'
import { ProcessableHandlerMixin } from '@livy/util/handlers/processable-handler-mixin'

/**
 * Stores log records in an array; great for debugging
 */
export class ArrayHandler
  extends MirrorSyncHandlerMixin(
    ProcessableHandlerMixin(AbstractLevelBubbleHandler),
  )
  implements SyncHandlerInterface, ResettableInterface
{
  private _records: LogRecord[] = []

  /**
   * @inheritdoc
   */
  public handleSync(record: LogRecord) {
    this._records.push(record)
    return !this.bubble
  }

  /**
   * @inheritdoc
   */
  public handleBatchSync(records: LogRecord[]) {
    this._records.push(...records)
  }

  /**
   * Get the stored records
   */
  public get records() {
    return this._records
  }

  /**
   * @inheritdoc
   */
  public reset() {
    this._records.splice(0)
  }
}
