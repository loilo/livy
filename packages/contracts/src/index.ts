export * from './closable-handler-interface.js'
export * from './formattable-handler-interface.js'
export * from './formatter-interface.js'
export * from './handler-interface.js'
export * from './log-level.js'
export * from './log-record.js'
export * from './logger-interface.js'
export * from './processable-handler-interface.js'
export * from './processor-interface.js'
export * from './resettable-interface.js'
