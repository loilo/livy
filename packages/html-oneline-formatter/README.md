# `@livy/html-oneline-formatter`

This [Livy](../../README.md#readme) formatter formats log records as single-line HTML.

## Basic Example

```js
import { FileHandler } from '@livy/file-handler'
import { HtmlOnelineFormatter } from '@livy/html-oneline-formatter'

const handler = new FileHandler('logs.html', {
  formatter: new HtmlOnelineFormatter()
})
```

## Installation

Install it via npm:

```bash
npm install @livy/html-oneline-formatter
```

## Options

An object of options can be passed to this formatter's constructor.

The following options are available:

### `include`

**Type:** [`Partial<IncludedRecordProperties>`](../util/README.md#includedrecordproperties)

**Default:** `{}`

**Description:** Which log record properties to include in the output. The passed object is merged into the following default:

```js
{
  datetime: true,
  channel: false,
  level: true,
  severity: false,
  message: true,
  context: true,
  extra: true
}
```

### `theme`

**Type:** [`HtmlFormatterThemeInterface`](src/themes/html-formatter-theme-interface.ts)

**Default:** `{}`

**Description:** The color theme to use. A couple of included themes can be found [in the themes folder](src/themes).

### `wrap`

**Type:** [`HtmlFormatterThemeInterface`](src/themes/html-formatter-theme-interface.ts)

**Default:** `false`

**Description:** Whether to allow lines to wrap when they're too long for the display.

## Public API

### `include`

Which log record properties to include in the output. Initially set through the `include` option.

### `theme`

The color theme to use. Initially set through the `theme` option.

### `wrap`

Whether to allow lines to wrap when they're too long for the display. Initially set through the `wrap` option.
