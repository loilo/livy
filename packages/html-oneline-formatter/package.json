{
  "name": "@livy/html-oneline-formatter",
  "type": "module",
  "version": "1.0.4",
  "description": "Formats Livy log records as single-line HTML",
  "keywords": [
    "logging",
    "logger",
    "logs",
    "log",
    "loglevel",
    "tools",
    "monolog",
    "formatter",
    "html"
  ],
  "author": "loilo <florian.reuschel@gmx.de>",
  "homepage": "https://github.com/loilo/livy/tree/master/packages/html-oneline-formatter#readme",
  "license": "MIT",
  "exports": {
    "types": "./lib/html-oneline-formatter.d.ts",
    "default": "./lib/html-oneline-formatter.js"
  },
  "sideEffects": false,
  "directories": {
    "lib": "lib",
    "test": "test"
  },
  "files": [
    "lib"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": "https://github.com/loilo/livy/tree/master/packages/html-oneline-formatter",
  "scripts": {
    "build": "del lib && tsc --declaration",
    "lint:readme": "remark --quiet --no-stdout -u validate-links README.md",
    "lint:source": "eslint \"src/**/*.ts\"",
    "test": "vitest run --coverage"
  },
  "bugs": {
    "url": "https://github.com/loilo/livy/issues"
  },
  "dependencies": {
    "@livy/contracts": "^1.1.2",
    "@livy/util": "^1.0.6",
    "@types/luxon": "^3.2.0",
    "@types/node": "^18.11.18",
    "hast-util-to-html": "^8.0.4",
    "highlight.js": "^11.7.0",
    "lowlight": "^2.8.0",
    "luxon": "^3.2.0",
    "unist-builder": "^3.0.0",
    "unist-util-map": "^3.1.2"
  },
  "devDependencies": {
    "typescript": "^5.1.0",
    "@livy/bin": "file:../../bin",
    "@livy/test-utils": "file:../../test-utils"
  },
  "gitHead": "bd96a3e4d1e5870873b0e0b9e538486aea6343a2"
}
