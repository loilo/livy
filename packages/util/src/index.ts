export type * from './formatters/abstract-batch-formatter.js'
export type * from './formatters/included-record-properties.js'
export type * from './formatters/line-formatter.js'
export type * from './handlers/abstract-batch-handler.js'
export type * from './handlers/abstract-formatting-processing-handler.js'
export type * from './handlers/abstract-level-bubble-handler.js'
export type * from './handlers/formattable-handler-mixin.js'
export type * from './handlers/is-closable-handler-interface.js'
export type * from './handlers/is-formattable-handler-interface.js'
export type * from './handlers/is-processable-handler-interface.js'
export type * from './handlers/is-sync-handler-interface.js'
export type * from './handlers/mirror-sync-handler-mixin.js'
export type * from './handlers/processable-handler-mixin.js'
export type * from './handlers/respect-level-mixin.js'
export type * from './console-adapter-interface.js'
export type * from './console-adapter.js'
export type * from './environment.js'
export type * from './gated-set.js'
export type * from './helpers.js'
export type * from './html.js'
export type * from './is-resettable-interface.js'
export type * from './mixin.js'
export type * from './timer.js'
export type * from './types.js'
export type * from './validatable-set.js'
