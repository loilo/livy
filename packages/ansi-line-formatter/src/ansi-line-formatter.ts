import type { LogLevel } from '@livy/contracts'
import { LineFormatter, LineFormatterOptions } from '@livy/line-formatter'
import { isEmpty } from '@livy/util/helpers'
import chalk, { supportsColor } from 'chalk'
import { DateTime } from 'luxon'

import { emphasize } from 'emphasize'
import chalkTemplate from 'chalk-template'

export interface AnsiLineFormatterOptions extends LineFormatterOptions {
  /**
   * Whether records should be highlighted with ANSI escape sequences
   * If not provided (or set to `undefined`), color support is going
   * to be automatically detected
   */
  decorated: boolean | undefined
}

/**
 * Formats log records as single lines with terminal highlighting
 */
export class AnsiLineFormatter extends LineFormatter {
  /**
   * Whether records should be highlighted with ANSI escape sequences
   */
  public decorated: boolean | undefined

  public constructor({
    decorated,
    ...options
  }: Partial<AnsiLineFormatterOptions> = {}) {
    super(options)
    this.decorated = decorated
  }

  /**
   * @inheritdoc
   */
  protected formatDatetime(datetime: DateTime) {
    const timestamp = super.formatDatetime(datetime)

    if (this.shouldDecorate()) {
      return chalk.dim(timestamp)
    } else {
      return timestamp
    }
  }

  /**
   * @inheritdoc
   */
  protected formatLevel(level: LogLevel) {
    const formatted = super.formatLevel(level)

    if (!this.shouldDecorate()) {
      return formatted
    } else {
      let color: string
      switch (level) {
        case 'emergency':
          color = 'red'
          break

        case 'alert':
          color = 'red'
          break

        case 'critical':
          color = 'red'
          break

        case 'error':
          color = 'red'
          break

        case 'warning':
          color = 'yellow'
          break

        case 'notice':
          color = 'blue'
          break

        case 'info':
          color = 'blue'
          break

        case 'debug':
          color = 'blue.dim'
          break

        /* c8 ignore next 2: This should never happen, but is included for type safety */
        default:
          return formatted
      }

      return chalkTemplate`{${color} ${formatted}}`
    }
  }

  /**
   * @inheritdoc
   */
  protected formatData(object: any, ignoreEmpty: boolean) {
    if (isEmpty(object) && ignoreEmpty) {
      return ''
    }

    const stringified = super.formatData(object)

    if (this.shouldDecorate()) {
      return emphasize.highlight('json', stringified).value
    } else {
      return stringified
    }
  }

  /**
   * Check whether the formatter should use ANSI codes to decorate log records
   */
  public shouldDecorate() {
    return Boolean(
      this.decorated !== undefined
        ? this.decorated
        : supportsColor
        ? supportsColor.hasBasic
        : false,
    )
  }
}
