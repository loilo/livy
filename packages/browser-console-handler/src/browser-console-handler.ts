import type { LogLevel, LogRecord, SyncHandlerInterface } from '@livy/contracts'
import { SeverityMap } from '@livy/contracts'
import {
  AbstractLevelBubbleHandler,
  AbstractLevelBubbleHandlerOptions,
} from '@livy/util/handlers/abstract-level-bubble-handler'
import { MirrorSyncHandlerMixin } from '@livy/util/handlers/mirror-sync-handler-mixin'
import { ProcessableHandlerMixin } from '@livy/util/handlers/processable-handler-mixin'
import { isEmpty } from '@livy/util/helpers'
import type { DateTime } from 'luxon'

export interface BrowserConsoleHandlerOptions
  extends AbstractLevelBubbleHandlerOptions {
  /**
   * Whether to use the browser's built-in console.debug() for the "debug" level,
   * which is only visible in the dev tools when explicitly configured so
   */
  useNativeDebug: boolean

  /**
   * Whether to include timestamps in the output
   */
  timestamps: boolean

  /**
   * The Console object to use
   */
  console: Console
}

/**
 * Writes log records to a browser console
 */
export class BrowserConsoleHandler
  extends MirrorSyncHandlerMixin(
    ProcessableHandlerMixin(AbstractLevelBubbleHandler),
  )
  implements SyncHandlerInterface
{
  private useNativeDebug: boolean
  private timestamps: boolean
  private console: Console

  public constructor({
    useNativeDebug = false,
    timestamps = false,
    console = self.console,
    ...options
  }: Partial<BrowserConsoleHandlerOptions> = {}) {
    super(options)

    this.console = console
    this.timestamps = timestamps
    this.useNativeDebug = useNativeDebug
  }

  /**
   * @inheritdoc
   */
  protected formatDatetime(datetime: DateTime) {
    return ['color: #888', datetime.toFormat('HH:mm:ss.SSS')]
  }

  /**
   * @inheritdoc
   */
  protected formatLevel(level: LogLevel) {
    const formatted = level.toUpperCase()

    let color: string
    switch (level) {
      case 'emergency':
        color = 'background-color: #f43a63; padding: 0 0.4em; color: #fff'
        break

      case 'alert':
        color = 'background-color: #f43a63; padding: 0 0.4em; color: #fff'
        break

      case 'critical':
        color = 'color: #f43a63'
        break

      case 'error':
        color = 'color: #f43a63'
        break

      case 'warning':
        color = 'color: #ffaa2b'
        break

      case 'notice':
        color = 'color: rgb(60, 161, 224)'
        break

      case 'info':
        color = 'color: rgb(60, 161, 224)'
        break

      case 'debug':
      default:
        color = 'color: rgba(60, 161, 224, 0.7)'
        break
    }

    return [color, formatted]
  }

  /**
   * @inheritdoc
   */
  public handleSync(record: LogRecord) {
    if (!this.isHandling(record.level)) {
      return
    }

    const resetter = 'color: inherit; background-color: inherit'
    const parameters: any[] = [record.message]
    const hasExtra = !isEmpty(record.extra)

    if (hasExtra || !isEmpty(record.context)) {
      parameters.push(record.context)
    }

    if (hasExtra) {
      parameters.push(record.extra)
    }

    const formattedTimestamp = this.timestamps
      ? [...this.formatDatetime(record.datetime), resetter]
      : []
    const formattedLevel = [...this.formatLevel(record.level), resetter]

    let placeholders = ['%c%s']
    if (this.timestamps) {
      placeholders.push('%c%s')
    }

    placeholders = placeholders.map(placeholder => `${placeholder}%c`)

    if (record.severity <= SeverityMap.error) {
      this.console.error(
        placeholders.join(' '),
        ...formattedTimestamp,
        ...formattedLevel,
        ...parameters,
      )
    } else if (record.severity === SeverityMap.warning) {
      this.console.warn(
        placeholders.join(' '),
        ...formattedTimestamp,
        ...formattedLevel,
        ...parameters,
      )
    } else if (record.severity <= SeverityMap.info) {
      this.console.info(
        placeholders.join(' '),
        ...formattedTimestamp,
        ...formattedLevel,
        ...parameters,
      )
    } else {
      if (this.useNativeDebug) {
        placeholders.pop()

        this.console.debug(
          placeholders.join(' '),
          ...formattedTimestamp,
          ...parameters,
        )
      } else {
        this.console.log(
          placeholders.join(' '),
          ...formattedTimestamp,
          ...formattedLevel,
          ...parameters,
        )
      }
    }

    return !this.bubble
  }
}
