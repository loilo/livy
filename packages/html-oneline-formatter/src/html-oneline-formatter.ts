import type { LogLevel, LogRecord, SeverityLevel } from '@livy/contracts'
import { AbstractBatchFormatter } from '@livy/util/formatters/abstract-batch-formatter'
import { IncludedRecordProperties } from '@livy/util/formatters/included-record-properties'
import { isEmpty, omit } from '@livy/util/helpers'
import { HTML } from '@livy/util/html'
import { Stringified } from '@livy/util/types'
import { toHtml as stringifyHtml } from 'hast-util-to-html'
import { lowlight } from 'lowlight'
import { Span, Text } from 'lowlight/lib/core.js'
import { DateTime } from 'luxon'
import { u } from 'unist-builder'
import { map } from 'unist-util-map'
import { HtmlFormatterThemeInterface } from './themes/html-formatter-theme-interface.js'

lowlight.registerLanguage(
  'json',
  (await import('highlight.js/lib/languages/json')).default,
)

/**
 * Relevant tokens of the lowlight parser
 */
const LowlightTokenMap = {
  attr: 'data_key',
  string: 'data_value_string',
  number: 'data_value_number',
  literal: 'data_value_literal',
} as const
// eslint-disable-next-line @typescript-eslint/no-redeclare
type LowlightTokenMap = typeof LowlightTokenMap

export interface HtmlOnelineFormatterOptions {
  /**
   * Which log record properties to include in the output
   */
  include: Partial<IncludedRecordProperties>

  /**
   * The color theme to use
   */
  theme: HtmlFormatterThemeInterface

  /**
   * Whether to allow lines to wrap when they're too long
   */
  wrap: boolean
}

/**
 * Formats log records as single-line HTML
 */
export class HtmlOnelineFormatter extends AbstractBatchFormatter {
  /**
   * @inheritdoc
   */
  protected batchDelimiter = ''

  /**
   * The color theme to use
   */
  public theme: HtmlFormatterThemeInterface

  /**
   * Whether to allow lines to wrap when they're too long
   */
  public wrap: boolean

  /**
   * Which log record properties to include in the output
   */
  public include: IncludedRecordProperties

  public constructor({
    theme = {},
    wrap = false,
    include = {},
  }: Partial<HtmlOnelineFormatterOptions> = {}) {
    super()

    this.theme = theme
    this.wrap = wrap
    this.include = {
      datetime: true,
      channel: false,
      level: true,
      severity: false,
      message: true,
      context: true,
      extra: true,
      ...include,
    }
  }

  /**
   * @inheritdoc
   */
  public format(record: LogRecord) {
    const message = record.message

    const formattedExtra = this.formatExtra(record.extra)

    return this.assembleFormattedRecord({
      datetime: this.formatDatetime(record.datetime),
      channel: this.formatChannel(record.channel),
      level: this.formatLevel(record.level),
      severity: this.formatSeverityMap(record.severity),
      message: this.formatMessage(message),
      context: this.formatContext(record.context, formattedExtra.length === 0),
      extra: formattedExtra,
    })
  }

  /**
   * Get the (CSS) color of a certain token
   *
   * @param token The token to get the color for
   */
  protected getColor(token: keyof HtmlFormatterThemeInterface) {
    if (token === 'background' && !('background' in this.theme)) {
      return 'transparent'
    }

    return this.theme[token] || 'inherit'
  }

  /**
   * Format a record's level
   *
   * @param level The level to format
   */
  protected formatLevel(level: LogLevel) {
    if (!this.include.level) {
      return ''
    }

    const color = this.getColor(
      `level_${level}` as keyof HtmlFormatterThemeInterface,
    )
    const text = level.toUpperCase()

    let punctuationPrefix = '['
    let punctuationSuffix = ']'

    const punctuationColor = this.getColor('punctuation')
    if (punctuationColor !== 'inherit') {
      punctuationPrefix = `<span style="color: ${punctuationColor}">[</span>`
      punctuationSuffix = `<span style="color: ${punctuationColor}">]</span>`
    }

    if (color !== 'inherit') {
      return `${punctuationPrefix}<span style="color: ${color}">${text}</span>${punctuationSuffix}`
    } else {
      return punctuationPrefix + text + punctuationSuffix
    }
  }

  /**
   * Format a record's severity
   *
   * @param severity The severity to format
   */
  protected formatSeverityMap(severity: SeverityLevel) {
    if (!this.include.severity) {
      return ''
    }

    const punctuationColor = this.getColor('punctuation')
    if (punctuationColor !== 'inherit') {
      return `<span style="color: ${punctuationColor}"> [${severity}]</span>`
    } else {
      return ` [${severity}]`
    }
  }

  /**
   * Format a record's channel
   *
   * @param channel The channel to format
   */
  protected formatChannel(channel: string) {
    if (!this.include.channel) {
      return ''
    }

    return channel
  }

  /**
   * Format a record's message
   *
   * @param message The message to format
   */
  protected formatMessage(message: string) {
    if (!this.include.message) {
      return ''
    }

    return message
  }

  /**
   * Get CSS style for a token
   *
   * @param token The context token to handle
   */
  protected getContextTokenStyle(
    token: keyof HtmlFormatterThemeInterface,
  ): string | undefined {
    const color = this.getColor(token)

    if (color !== 'inherit') {
      return `color: ${color}`
    } else {
      return undefined
    }
  }

  /**
   * Serialize the style of a log record as CSS
   */
  protected stringifyStyles() {
    const style = {
      margin: '0',
      padding: '0.3em 0.8em 0.3em',
      fontFamily:
        "'SFMono-Regular', Consolas, 'Liberation Mono', Menlo, Courier, monospace",
      whiteSpace: this.wrap ? 'pre-wrap' : 'nowrap',
      backgroundColor: this.getColor('background'),
      color: this.getColor('text'),
    }

    return Object.entries(style)
      .filter(([, value]) => typeof value === 'string' && value.length > 0)
      .map(
        ([property, value]) =>
          `${property.replace(
            /([A-Z])/,
            (_, letter) => `-${letter.toLowerCase()}`,
          )}: ${value}`,
      )
      .join(';')
  }

  /**
   * Assemble the formatted parts of a record into a string
   *
   * @param parts The formatted parts to assemble
   */
  protected assembleFormattedRecord({
    channel,
    datetime,
    level,
    severity,
    message,
    context,
    extra,
  }: Stringified<LogRecord>) {
    return `<pre style="${this.stringifyStyles()}">${datetime}${channel} ${level}${severity} ${message}${context}${extra}</pre>`
  }

  /**
   * Serialize a piece of data
   *
   * @param data The data to serialize
   */
  protected serializeData(data: any) {
    const result = JSON.stringify(data)

    return typeof result !== 'string' ? 'undefined' : result
  }

  /**
   * Format the log record's time
   *
   * @param datetime The DateTime object to format
   */
  protected formatDatetime(datetime: DateTime) {
    if (!this.include.datetime) {
      return ''
    }

    const punctuationColor = this.getColor('punctuation')
    const formattedTime = datetime.toFormat('yyyy-MM-dd HH:mm:ss ')

    if (punctuationColor !== 'inherit') {
      return HTML`<span style="color: ${this.getColor(
        'punctuation',
      )}">${formattedTime}</span>`.toString()
    } else {
      return formattedTime
    }
  }

  /**
   * Format a record's context object
   *
   * @param context     The context to format
   * @param ignoreEmpty Whether to return an empty serialization for an empty context object
   */
  protected formatContext(context?: any, ignoreEmpty = true) {
    if (!this.include.context) {
      return ''
    }

    const formatted = this.formatData(context, ignoreEmpty)

    if (formatted.length === 0 && ignoreEmpty) {
      return ''
    } else {
      return ` ${formatted}`
    }
  }

  /**
   * Format a record's extra object
   *
   * @param context     The extra to format
   * @param ignoreEmpty Whether to return an empty serialization for an empty extra object
   */
  protected formatExtra(extra?: any, ignoreEmpty = true) {
    if (!this.include.extra) {
      return ''
    }

    const formatted = this.formatData(extra, ignoreEmpty)
    if (formatted.length === 0 && ignoreEmpty) {
      return ''
    } else {
      return ` ${formatted}`
    }
  }

  /**
   * Format a single record property
   *
   * @param data        The data to format
   * @param ignoreEmpty Whether to return an empty serialization for empty data
   */
  protected formatData(data: any, ignoreEmpty = true) {
    if (ignoreEmpty && isEmpty(data)) {
      return ''
    } else {
      const stringified = this.serializeData(data)

      if (
        !('data_key' in this.theme) &&
        !('data_value_string' in this.theme) &&
        !('data_value_number' in this.theme) &&
        !('data_value_literal' in this.theme)
      ) {
        return stringified
      }

      type StyleSpan = Omit<Span, 'properties'> & {
        properties: Omit<Span['properties'], 'className'> & {
          style: string
        }
      }

      const tree = lowlight.highlight('json', stringified).children as (
        | Text
        | Span
        | StyleSpan
      )[]

      const highlightedContextTree = map(u('root', tree), node => {
        if (
          node &&
          node.type === 'element' &&
          typeof node.properties === 'object' &&
          'className' in node.properties
        ) {
          return {
            ...node,
            properties: {
              ...omit(node.properties, 'className'),
              style: node.properties.className
                .filter((className: string) => className.startsWith('hljs-'))
                .map((className: string) =>
                  this.getContextTokenStyle(
                    LowlightTokenMap[
                      className.slice(5) as keyof LowlightTokenMap
                    ],
                  ),
                )
                .filter(Boolean)
                .join('; '),
            },
          }
        } else {
          return node
        }
      })

      const formattedJson = stringifyHtml(highlightedContextTree).toString()

      const punctuationColor = this.getColor('punctuation')

      if (punctuationColor !== 'inherit') {
        return `<span style="color: ${punctuationColor}">${formattedJson}</span>`
      } else {
        return formattedJson
      }
    }
  }
}
