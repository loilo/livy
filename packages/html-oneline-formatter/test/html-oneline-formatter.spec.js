import { describe, expect, it } from 'vitest'
import { HtmlOnelineFormatter } from '../src/html-oneline-formatter'

const { record } = livyTestGlobals

describe('@livy/html-oneline-formatter', () => {
  it('should include all parts except channel by default', () => {
    const htmlOnelineFormatter = new HtmlOnelineFormatter()
    expect(
      htmlOnelineFormatter.format(
        record('info', 'Test HtmlOnelineFormatter', {
          context: { foo: 1 },
          extra: { bar: 2 }
        })
      )
    ).toMatchSnapshot()
  })

  it('should join batched formats with the empty string', () => {
    const htmlOnelineFormatter = new HtmlOnelineFormatter()
    expect(
      htmlOnelineFormatter.formatBatch([
        record('info', 'Test HtmlOnelineFormatter', {
          context: { foo: 1 },
          extra: { bar: 2 }
        }),
        record('info', 'Test HtmlOnelineFormatter', {
          context: { foo: 1 },
          extra: { bar: 2 }
        })
      ])
    ).toMatchSnapshot()
  })

  it('should omit the "extra" part completely if it\'s empty', () => {
    const htmlOnelineFormatter = new HtmlOnelineFormatter()
    expect(
      htmlOnelineFormatter.format(record('info', 'Test HtmlOnelineFormatter'))
    ).toMatchSnapshot()
  })

  it('should respect the "include" option', () => {
    const htmlOnelineFormatter = new HtmlOnelineFormatter({
      include: {
        channel: true,
        severity: true
      }
    })

    expect(htmlOnelineFormatter.include).toEqual({
      datetime: true,
      channel: true,
      level: true,
      severity: true,
      message: true,
      context: true,
      extra: true
    })

    expect(
      htmlOnelineFormatter.format(record('info', 'Test HtmlOnelineFormatter'))
    ).toMatchSnapshot()
  })

  it('should return an empty line when no properties are included', () => {
    const htmlOnelineFormatter = new HtmlOnelineFormatter({
      include: {
        datetime: false,
        channel: false,
        level: false,
        severity: false,
        message: false,
        context: false,
        extra: false
      }
    })

    expect(
      htmlOnelineFormatter.format(record('info', 'Test HtmlOnelineFormatter'))
    ).toMatch(/<pre[^>]*>\s*<\/pre>/)
  })

  it('should respect the "wrap" option', () => {
    const htmlOnelineFormatter = new HtmlOnelineFormatter({
      wrap: true
    })

    expect(
      htmlOnelineFormatter.format(record('info', 'Test HtmlOnelineFormatter'))
    ).toMatchSnapshot()
  })

  it('should respect the "theme" option', () => {
    const htmlOnelineFormatter = new HtmlOnelineFormatter({
      include: {
        severity: true
      },
      theme: {
        text: 'black',
        punctuation: 'gray',
        level_emergency: 'darkred',
        level_alert: 'crimson',
        level_critical: 'indianred',
        level_error: 'lightcoral',
        level_warning: 'gold',
        level_notice: 'steelblue',
        level_info: 'dodgerblue',
        level_debug: 'lightskyblue',
        data_key: 'inherit',
        data_value_string: 'mediumseagreen',
        data_value_number: 'darkslateblue',
        data_value_literal: 'sandybrown'
      }
    })

    expect(
      htmlOnelineFormatter.format(
        record('info', 'Test HtmlOnelineFormatter', {
          context: { foo: 1 },
          extra: { bar: 2 }
        })
      )
    ).toMatchSnapshot()
  })

  it('should omit punctuation wrapper elements if punctuation color is "inherit"', () => {
    const htmlOnelineFormatter = new HtmlOnelineFormatter({
      theme: {
        text: 'black',
        punctuation: 'inherit',
        level_emergency: 'darkred',
        level_alert: 'crimson',
        level_critical: 'indianred',
        level_error: 'lightcoral',
        level_warning: 'gold',
        level_notice: 'steelblue',
        level_info: 'dodgerblue',
        level_debug: 'lightskyblue',
        data_key: 'inherit',
        data_value_string: 'mediumseagreen',
        data_value_number: 'darkslateblue',
        data_value_literal: 'sandybrown'
      }
    })

    expect(
      htmlOnelineFormatter.format(
        record('info', 'Test HtmlOnelineFormatter', {
          context: { foo: 1 },
          extra: { bar: 2 }
        })
      )
    ).toMatchSnapshot()
  })
})
