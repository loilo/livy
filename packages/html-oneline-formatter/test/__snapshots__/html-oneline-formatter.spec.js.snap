// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@livy/html-oneline-formatter > should include all parts except channel by default 1`] = `"<pre style=\\"margin: 0;padding: 0.3em 0.8em 0.3em;font-family: 'SFMono-Regular', Consolas, 'Liberation Mono', Menlo, Courier, monospace;white-space: nowrap;background-color: transparent;color: inherit\\">2015-08-15 14:30:00  [INFO] Test HtmlOnelineFormatter {\\"foo\\":1} {\\"bar\\":2}</pre>"`;

exports[`@livy/html-oneline-formatter > should join batched formats with the empty string 1`] = `"<pre style=\\"margin: 0;padding: 0.3em 0.8em 0.3em;font-family: 'SFMono-Regular', Consolas, 'Liberation Mono', Menlo, Courier, monospace;white-space: nowrap;background-color: transparent;color: inherit\\">2015-08-15 14:30:00  [INFO] Test HtmlOnelineFormatter {\\"foo\\":1} {\\"bar\\":2}</pre><pre style=\\"margin: 0;padding: 0.3em 0.8em 0.3em;font-family: 'SFMono-Regular', Consolas, 'Liberation Mono', Menlo, Courier, monospace;white-space: nowrap;background-color: transparent;color: inherit\\">2015-08-15 14:30:00  [INFO] Test HtmlOnelineFormatter {\\"foo\\":1} {\\"bar\\":2}</pre>"`;

exports[`@livy/html-oneline-formatter > should omit punctuation wrapper elements if punctuation color is "inherit" 1`] = `"<pre style=\\"margin: 0;padding: 0.3em 0.8em 0.3em;font-family: 'SFMono-Regular', Consolas, 'Liberation Mono', Menlo, Courier, monospace;white-space: nowrap;background-color: transparent;color: black\\">2015-08-15 14:30:00  [<span style=\\"color: dodgerblue\\">INFO</span>] Test HtmlOnelineFormatter <span style=\\"\\">{</span><span style=\\"\\">\\"foo\\"</span><span style=\\"\\">:</span><span style=\\"color: darkslateblue\\">1</span><span style=\\"\\">}</span> <span style=\\"\\">{</span><span style=\\"\\">\\"bar\\"</span><span style=\\"\\">:</span><span style=\\"color: darkslateblue\\">2</span><span style=\\"\\">}</span></pre>"`;

exports[`@livy/html-oneline-formatter > should omit the "extra" part completely if it's empty 1`] = `"<pre style=\\"margin: 0;padding: 0.3em 0.8em 0.3em;font-family: 'SFMono-Regular', Consolas, 'Liberation Mono', Menlo, Courier, monospace;white-space: nowrap;background-color: transparent;color: inherit\\">2015-08-15 14:30:00  [INFO] Test HtmlOnelineFormatter</pre>"`;

exports[`@livy/html-oneline-formatter > should respect the "include" option 1`] = `"<pre style=\\"margin: 0;padding: 0.3em 0.8em 0.3em;font-family: 'SFMono-Regular', Consolas, 'Liberation Mono', Menlo, Courier, monospace;white-space: nowrap;background-color: transparent;color: inherit\\">2015-08-15 14:30:00 logs [INFO] [6] Test HtmlOnelineFormatter</pre>"`;

exports[`@livy/html-oneline-formatter > should respect the "theme" option 1`] = `"<pre style=\\"margin: 0;padding: 0.3em 0.8em 0.3em;font-family: 'SFMono-Regular', Consolas, 'Liberation Mono', Menlo, Courier, monospace;white-space: nowrap;background-color: transparent;color: black\\"><span style=\\"color: gray\\">2015-08-15 14:30:00 </span> <span style=\\"color: gray\\">[</span><span style=\\"color: dodgerblue\\">INFO</span><span style=\\"color: gray\\">]</span><span style=\\"color: gray\\"> [6]</span> Test HtmlOnelineFormatter <span style=\\"color: gray\\"><span style=\\"\\">{</span><span style=\\"\\">\\"foo\\"</span><span style=\\"\\">:</span><span style=\\"color: darkslateblue\\">1</span><span style=\\"\\">}</span></span> <span style=\\"color: gray\\"><span style=\\"\\">{</span><span style=\\"\\">\\"bar\\"</span><span style=\\"\\">:</span><span style=\\"color: darkslateblue\\">2</span><span style=\\"\\">}</span></span></pre>"`;

exports[`@livy/html-oneline-formatter > should respect the "wrap" option 1`] = `"<pre style=\\"margin: 0;padding: 0.3em 0.8em 0.3em;font-family: 'SFMono-Regular', Consolas, 'Liberation Mono', Menlo, Courier, monospace;white-space: pre-wrap;background-color: transparent;color: inherit\\">2015-08-15 14:30:00  [INFO] Test HtmlOnelineFormatter</pre>"`;
